### PR TITLE
More words on Decoupled and Irrevocable

### PR DESCRIPTION
**Pull request by schoeberl**

### CODE REVIEW

File: docs/src/main/tut/chisel3/interfaces-and-connections.md [hunk 1]
```diff
@@ -129,3 +132,14 @@ class ConsumingData extends Module {
...
 }
 ```
+
+Decoupled is marker interface for a ready-valid interface to indicate
+that there are no requirements placed on the signaling of `ready` or `valid`.
+That means `ready` and `valid` can also be de-asserted without a data transfer.
+
+Irrevocable is marker interface for a ready-valid interface to indicate
+that the value of 'bits' after a cycle where 'valid' is high and 'ready' is low
+are not changed. E.g., `valid` stays asserted. Also the consumer shall
```
Comment (seldridge) [inline docs/src/main/tut/chisel3/interfaces-and-connections.md]:
```suggestion
Also the consumer shall
```

File: docs/src/main/tut/chisel3/interfaces-and-connections.md [hunk 2]
```diff
@@ -129,3 +132,14 @@ class ConsumingData extends Module {
...
+keep `ready` asserted after a cycle where `read` was high and `valid` was low.
```
Comment (seldridge) [inline docs/src/main/tut/chisel3/interfaces-and-connections.md]: I think this is wrong. The convention is that `valid` will stay high and not de-assert. I don't think there's any guarantees on `ready`?

File: docs/src/main/tut/chisel3/interfaces-and-connections.md [hunk 3]
```diff
@@ -129,3 +132,14 @@ class ConsumingData extends Module {
...
+Note that this constraint cannot be enforced by the interface.
```
Comment (seldridge) [inline docs/src/main/tut/chisel3/interfaces-and-connections.md]:
```suggestion
Note that the *irrevocable* constraint *is only a convention* and cannot be enforced by the interface.
```

File: docs/src/main/tut/chisel3/interfaces-and-connections.md [hunk 4]
```diff
@@ -83,6 +83,9 @@ Caution: bulk connections should only be used with **directioned elements** (lik
 ## The standard ready-valid interface (ReadyValidIO / Decoupled)
 
 Chisel provides a standard interface for [ready-valid interfaces](http://inst.eecs.berkeley.edu/~cs150/Documents/Interfaces.pdf).
+The ready-valid interface uses `ready` and `valid` to handshake
+data transfer. When both signals are asserted the data transfer takes
+place.
```
Comment (seldridge) [inline docs/src/main/tut/chisel3/interfaces-and-connections.md]: I don't like `"handshake"` being used as a verb here. It also seems reasonable to be explicit about what each signal means and to document `fire`.

How about:

```markdown
A ready-valid interface consists of a `ready` bit, a `valid` bit, and some data stored in `bits`.
The `ready` bit indicates that a consumer is *ready* to consume data.
The `valid` bit indicates that a produced has *valid* data on `bits`.
When both `ready` and `valid` are asserted, a data transfer from producer to consumer takes place.
A convenience method `fire` is provided that is asserted if both `ready` and `valid` are asserted.
```

(GitHub really, really needs multi-line suggestions...)

**Comment by seldridge** [comment]
Also, website build is busted due to my own fault and nothing to do with this PR. I'll fix this (hopefully) later today.

**Comment by schoeberl** [comment]
> On 23 Sep, 2019, at 19:15, Schuyler Eldridge <notifications@github.com> wrote:
> 
> @seldridge requested changes on this pull request.
> 
> Thanks for diving into the website here, @schoeberl <https://github.com/schoeberl>! Both decoupled and irrevocable need more documentation.
> 
> I have some comments, most of which I hope are constructive and not wordsmithing.
> 
Yes, thanks for your feedback :-)

I tried to incorporate it with some tweaks.

> 
> 
> +Irrevocable is marker interface for a ready-valid interface to indicate
> +that the value of 'bits' after a cycle where 'valid' is high and 'ready' is low
> +are not changed. E.g., `valid` stays asserted. Also the consumer shall
> +keep `ready` asserted after a cycle where `read` was high and `valid` was low.
> I think this is wrong. The convention is that valid will stay high and not de-assert. I don't think there's any guarantees on ready?
> 

There is no wrong or right here, as this is a convention. In ready-valid there are four convention options
after one ready xor valid assertion in the next clock cycle

1.) Both can change as they like (DecoupledIO)
2.) Both need to stay asserted if they have been before (which is what I think Irrevocable is)
3.) Valid/bits need to stay, ready may change (your interpretation)
4.) Ready needs to stay, valid/ready may change (not covered)

AXI is 2.) when a master or slave has decided on valid or ready it is not allowed to change its mind.

Cheers,
Martin

**Comment by schoeberl** [comment]
I came up with a 5th version: asa  producer you are not allowed to deassert valid after being asserted,
but you may change the data. Imagine an always valid temperature sensor that changes
the value independent of being read or not.

BTW, “data" would have been a better name than “bits" for the interface/bundle.

Cheers,
Martin

File: docs/src/main/tut/chisel3/interfaces-and-connections.md [hunk 5]
```diff
@@ -129,3 +137,18 @@ class ConsumingData extends Module {
...
+That means `ready` and `valid` can also be de-asserted without a data transfer.
+Specifically, `ready` and `valid` may be deasserted and `bits` may change
+after a clock cycle without a data transfer.
+
+`IrrevocableIO` is a ready-valid interface with the *convention*
+that the value of `bits` will not change while `valid` is asserted and `ready` is deasserted.
+Also the consumer shall keep `ready` asserted after a cycle where `read`
+was high and `valid` was low.
+Note that the *irrevocable* constraint *is only a convention* and cannot
+be enforced by the interface.
```
Comment (jackkoenig) [inline docs/src/main/tut/chisel3/interfaces-and-connections.md]: > Note that the *irrevocable* constraint *is only a convention* and cannot be enforced by the interface.

This statement is true of most interfaces (including AXI and TileLink). What I would note is that we don't generate checkers/assertions.

File: docs/src/main/tut/chisel3/interfaces-and-connections.md [hunk 6]
```diff
@@ -129,3 +132,14 @@ class ConsumingData extends Module {
   // do something with io.readyValid.bits
 }
 ```
+
+Decoupled is marker interface for a ready-valid interface to indicate
+that there are no requirements placed on the signaling of `ready` or `valid`.
+That means `ready` and `valid` can also be de-asserted without a data transfer.
```
[the remark this answers is not in the document]
Comment (schoeberl) [reply docs/src/main/tut/chisel3/interfaces-and-connections.md]: I have "Specifically, `ready` and `valid` may be deasserted and `bits` may change
after a clock cycle without a data transfer."

I think "bits may change while valid asserted" is not enough. They can change with valid or without valid.

**Pull request closed without merging**